thread_pool.h

```diff
@@ -25,35 +30,97 @@ class ThreadPool {
     unsigned num_active_threads = 0;
     //FIXME temporary impl
     std::vector<std::thread> threads;
+    bool stoping = false;
+    std::condition_variable event_var;
+    std::mutex event_mutex;
+    std::queue <std::function<void()>>  task_queue;
 
 public:
-    explicit ThreadPool(ThreadPool::Config config = {}) {} //TODO implement
+    explicit ThreadPool(ThreadPool::Config config = {}) {
+        start();
+    } //TODO implement
 
     ~ThreadPool() {
-        //FIXME temporary impl
-        for (auto& th : threads) {
-            th.join();
-        }
+        stop();
         //close active connections
         //reject pending connections
         //send shutdown signal
         //wait for threads destruction
     }
 
-    bool submit_job(std::function<void()>&& job) {
+    auto submit_job(std::function<void()>&& job)->std::future<decltype(job())> 
+    {
+        
         //FIXME temporary impl
-        threads.emplace_back(std::forward<std::function<void()>>(job));
+        //threads.emplace_back(std::forward<std::function<void()>>(job));
         //if accepted
-        return true;
+        if (task_queue::size() < Config::task_queue_size) {
+
+            auto wrapper = std::make_shared<std::packaged_task<decltype(job()) ()>>(std::move(job()));
```

Comment:
- Why do you need to use `decltype`? The return type of the functor is known -> void
- This isn't asynchronous execution. You are calling the `job` object and moving its result. This will be either a compile time error (since void doesn't define operator()) or a runtime error because `packaged_task` expects a callable object (like `function`).
- Why do you need a pointer to the packaged task?